ui-manchette/src/components/helpers.ts

```diff
@@ -1,3 +1,97 @@
-export const positionMmtoKm = (position: number) => {
+import type { OperationalPoint } from '@osrd-project/ui-spacetimechart/dist/lib/types';
+import type { OperationalPointType, StyledOperationalPointType } from '../types';
+import { BASE_KM_HEIGHT, BASE_OP_HEIGHT } from './consts';
+
+export const positionMmToKm = (position: number) => {
   return Math.round((position / 1000000) * 10) / 10;
 };
+
+export const calcOperationalPointsToDisplay = (
+  operationalPoints: OperationalPointType[],
+  isProportionnal: boolean,
+  zoom: number
+) => {
+  if (isProportionnal) {
+    const result = operationalPoints.reduce(
+      (acc, nextOp) => {
+        let op = { ...acc[acc.length - 1] };
+        const newOp = { ...nextOp, display: true, styles: {} };
+        const diff = positionMmToKm(nextOp.position - op.position);
+        if (diff * 8 * zoom >= BASE_OP_HEIGHT) {
+          acc.push(newOp);
+          op = nextOp;
+        } else {
+          newOp.display = false;
+          acc.push(newOp);
+        }
+        return acc;
+      },
+      [{ ...operationalPoints[0], display: true, styles: {} } as StyledOperationalPointType]
+    );
+
+    const lastElement = operationalPoints[operationalPoints.length - 1];
+    if (result[result.length - 1] !== lastElement) {
+      result.push(lastElement);
+    }
+
+    const secondLastIndex = result.length - 2;
+    if (secondLastIndex >= 0) {
+      const secondLastElement = result[secondLastIndex];
+      const lastDiff = positionMmToKm(lastElement.position - secondLastElement.position);
+      if (lastDiff * 8 * zoom < BASE_OP_HEIGHT) {
+        result.splice(secondLastIndex, 1);
+      }
+    }
+    result.shift();
+    return result;
+  } else {
+    return operationalPoints as StyledOperationalPointType[];
+  }
+};
+
+export const operationalPointsHeight = (
```

Comment:
I see that in this function, you mutate existing `styles` attributes for the input `operationalPoints`.

I think this pattern is dangerous: When I see this function that takes a `StyledOperationalPointType[]` and returns another `StyledOperationalPointType[]`, I would expect it not to mutate what's inside the input value:
- Either you return something totally new
- Either you mutate, and name and type the function accordingly (like `const setOperationalPointsHeight: (...) => void` for instance)

I know JavaScript itself does not respect this convention (like with `Array#reverse` for instance), but it has been widely criticized, and it is now quite standard to have no side effects on functions that return something new.

Anyway, a common way to avoid mutating the input would be to replace the `op.styles = {...}; return op;` by `return {...op, styles: {...}};`.